source/extensions/access_loggers/http_grpc/grpc_access_log_impl.h

```diff
@@ -90,27 +70,39 @@ class GrpcAccessLogStreamerImpl : public Singleton::Instance, public GrpcAccessL
     void onReceiveTrailingMetadata(Http::HeaderMapPtr&&) override {}
     void onRemoteClose(Grpc::Status::GrpcStatus status, const std::string& message) override;
 
-    ThreadLocalStreamer& parent_;
-    const std::string log_name_;
+    GrpcAccessLoggerImpl& parent_;
     Grpc::AsyncStream<envoy::service::accesslog::v2::StreamAccessLogsMessage> stream_{};
   };
 
+  Grpc::AsyncClient<envoy::service::accesslog::v2::StreamAccessLogsMessage,
+                    envoy::service::accesslog::v2::StreamAccessLogsResponse>
+      client_;
+  std::string log_name_;
+  absl::optional<LocalStream> stream_;
+  const LocalInfo::LocalInfo& local_info_;
+};
+
+class GrpcAccessLoggerCacheImpl : public Singleton::Instance, public GrpcAccessLoggerCache {
+public:
+  GrpcAccessLoggerCacheImpl(Grpc::AsyncClientManager& async_client_manager, Stats::Scope& scope,
+                            ThreadLocal::SlotAllocator& tls,
+                            const LocalInfo::LocalInfo& local_info);
+
+  GrpcAccessLoggerSharedPtr getOrCreateLogger(
+      const ::envoy::config::accesslog::v2::CommonGrpcAccessLogConfig& config) override;
+
+private:
   /**
-   * Per-thread multi-stream state.
+   * Per-thread cache.
    */
-  struct ThreadLocalStreamer : public ThreadLocal::ThreadLocalObject {
-    ThreadLocalStreamer(const SharedStateSharedPtr& shared_state);
-    void send(envoy::service::accesslog::v2::StreamAccessLogsMessage& message,
-              const std::string& log_name);
-
-    Grpc::AsyncClient<envoy::service::accesslog::v2::StreamAccessLogsMessage,
-                      envoy::service::accesslog::v2::StreamAccessLogsResponse>
-        client_;
-    std::unordered_map<std::string, ThreadLocalStream> stream_map_;
-    SharedStateSharedPtr shared_state_;
+  struct ThreadLocalCache : public ThreadLocal::ThreadLocalObject {
+    std::unordered_map<std::size_t, GrpcAccessLoggerSharedPtr> access_loggers_;
```

Comment:
nit: use absl_flat_hash_map, also please add a comment saying what the size_t param is.